src/clarity_cli.rs

```diff
@@ -1795,29 +1721,6 @@ pub fn invoke_command(invoked_by: &str, args: &[String]) -> (i32, Option<serde_j
                 }
             }
         }
-        "make_lcov" => {
```

Comment:
Maybe worth noting for other reviewers that the code coverage is being removed from that codebase because it's considered as a developer tool and the release cycle of such a tool should probably not be correlated with releases of the Stacks blockchain.
Also, we iterated on the initial approach produced here by Aaron in Clarinet, we have some additional informations being extracted and we will keep improving it in the future with other information like branch coverage - not covered today, which could lead developers thinking that their code is tested when it's not if multiple statements are present on the same line - which could be problematic for a tool coming from that codebase.

Reply:
Curious to hear that you're in favor of keeping code :).

> it already works and is tested

I could be missing something, but could you point us to the tests?

> The question we should be asking is whether or not the loss from breaking downstream projects by dropping this feature is worth it [...] But I can tell you that there are projects (i.e. mine) that rely on this feature

We of course considered this, but in all honesty, how many times did you use this feature - knowing that per https://github.com/stacks-network/stacks-blockchain/issues/3160,  it's been broken between June 15 2021 and fixed on June 6 2022 by @hstove in https://github.com/stacks-network/stacks-blockchain/issues/3160, who migrated to clarinet a few weeks later?

Reply:
My thought here was just that this was a logical simplification to make the core code, since it is not needed for the node, doesn't really feel like it belongs in here, and can now be easily implemented by other tools.

I am okay with putting it back if you like.

Reply:
> **most Clarity code already adheres to this**, including the code samples we use in the documentation, the Clarity book, and tutorials.

Also, I don't think I've never heard of this being a good practice or a recommendation, and the Clarity book - let's randomly take this page for instance [clarity book](https://book.clarity-lang.org/ch05-01-public-functions.html)) - is definitely not pushing for this.

Reply:
@lgalabru lcov was added by #2592, and Hank reported the breakage on June 6, 2022.  I'm not sure from his report which PR he was referencing that specifically removed it, but #3163 restored it.  I had built and kept a `clarity-cli` binary outside of the source tree that had working lcov which I used for my tests, so I didn't notice the breakage in the code right away.

> We of course considered this,

Who is "we"?

>  doesn't really feel like it belongs in here, and can now be easily implemented by other tools.

Normally, I would agree if the "other tools" (i.e. Clarinet) were usable on the same target systems as `clarity-cli`.  As of today, this is not the case.  Specifically, `deno` does not build on `x86_64-unknown-linux-musl` (see https://github.com/denoland/rusty_v8/issues/49).  So until this is fixed, I think it's important that we retain `clarity-cli`'s feature set to the greatest extent possible, even if it overlaps in scope with Clarinet and other tools.

Reply:
"We" is Ludo and me; we discussed this earlier.

Reply:
> Also, I don't think I've never heard of this being a good practice or a recommendation, and the Clarity book - let's randomly take this page for instance clarity book) - is definitely not pushing for this.

Here is the example you linked: 

```clarity
(define-data-var even-values uint u0)

(define-public (count-even (number uint))
    (begin
        ;; increment the "event-values" variable by one.
        (var-set even-values (+ (var-get even-values) u1))
        
        ;; check if the input number is even (number mod 2 equals 0).
        (if (is-eq (mod number u2) u0)
            (ok "the number is even")
            (err "the number is uneven")
        )
    )
)

;; Call count-even two times.
(print (count-even u4))
(print (count-even u7))

;; Will this return u1 or u2?
(print (var-get even-values))
```

What would happen if I ran lcov on unit test for `(count-even u1)`?  Would lcov falsely report that some testable code in `count-even` is being reached, when it is not?  Would lcov falsely report that some testable code was not being reached, when it actually was?  If the answer to both of these is "no," then wouldn't it be fair to say that lcov is providing some value to the developer?

Reply:
> I'm not sure from his report which PR he was referencing that specifically removed it, [...] I had built and kept a clarity-cli binary outside of the source tree that had working lcov which I used for my tests, so I didn't notice the breakage in the code right away.

The feature was broken since before its initial merge in `develop`. If you look at the final diffs https://github.com/stacks-network/stacks-blockchain/pull/2592/files, the feature was implemented but the actual CLI integration was implemented and reverted in the same PR. So if you had a working copy of `clarinet-cli`, it was coming from the early commits of that branch.
Hank is the only developer who noticed that issue, after one year.

> What would happen if [...]

My point was that we're not educating developers for 1 statement, 1 line as you said earlier, but let's get through it. 
In the short snippet that you've captured, it's all fine. If you take the few lines located after that snippet, on that same page:

```clarity
(asserts! (is-eq (get-counter-of 'ST1J4G6RR643BCG8G8SR6M2D9Z9KXT2NJDRK3FBTK) u5) 
(asserts! (is-eq (get-counter-of 'ST20ATRN26N9P05V2F1RHFRV24X8C8M3W54E427B2) u10) 
(asserts! (is-eq (get-counter-of 'ST21HMSJATHZ888PD0S0SSTWP4J61TCRJYEVQ0STB) u0) 
```
This pattern and style are pretty typical, you can also find them in your own [contracts](https://github.com/jcnelson/poxl/blob/main/tests/test-poxl.clar). In this context, not having branch coverage becomes more problematic because we're indicating that the early exit statements are covered by unit tests, when they are not. In other words: "no need to write more tests for this function, it will be fine".  

> providing some value to the developer

I'm not saying that the current version is not providing value, even without branch coverage the value is immense. 
My point is that code coverage need to receive more updates in the future, the current version is imperfect and can be misleading in certain cases (that are not edge cases). 

> clarinet does not build on `x86_64-unknown-linux-musl`

Surprising, that could be the case, I personally never tried it and I don't think we received a complaint about this in our repo which explains that this limitation was definitely not in our radar. 
I just created an issue for tracking this problem [here](https://github.com/hirosystems/clarinet/issues/518), if you have some logs that could be helpful, thanks!

Reply:
Thank you for your thorough reply.

We agree that the current lcov support is helpful but is broken.  But, it's surmountable.  If you look at some of my subsequent [contracts](https://github.com/jcnelson/nftree/blob/main/contracts/nftree.clar), you'll see that I have since switched my style to putting statements in early-returns on separate lines.

> My point is that code coverage need to receive more updates in the future, the current version is imperfect and can be misleading in certain cases (that are not edge cases).

While this is all true, I don't think that we can conclude that we should remove lcov.  I think we should instead strive to make it better.  If users know what the pitfalls are in advance (e.g. the lcov switch can print them), then I'm not seeing how removing a value-adding feature is improving the code here?

Reply:
Switching `clarity-cli` over to the coverage implementation from Clarinet requires some deeper changes to `clarity-cli` that I didn't want to get into for the purposes of this PR. Instead, I decided to just put back the existing coverage implementation, and modify it to use `EvalHook`. That is now complete with bcec67109fb0a5afba78025914ea80057bd0456b.

Reply:
Sounds good :+1: